src/api_sketch.jl

```diff
@@ -0,0 +1,211 @@
+that will essentially copy the current contents of atomfeat.jl, ideally with some
+additions enabling the user to augment the lookup table
+
+should also have convenience functions for building encode_f and decode_f via keywords for, e.g.:
+    choosing one-hot (and how many bins, etc.) vs. direct float encoding
+    logspaced vs. linear spaced binning for one-hot encoding
+    maybe other stuff too
+...
+=#
+
+# then basically copy a bunch of the weave stuff, plus potentially things like oxidation state etc. from pymatgen belong here too
+
+#=
+Feature of a pair of atoms. Currently only used in WeaveModel, but may eventually
+have a version of AtomGraph that allows edge features...
+=#
+struct PairFeat{Tn,Te}<:AbstractFeature{Tn,Te}
+    name
+    encode_f
+    decode_f
+    length::Int # maybe, maybe not (does constrain/assume vector Te)
+    # probably needs some other stuff...
+end
+# same idea here...copy from weave stuff...
+# a note: I think it's more elegant (and probably efficient?) if pair features assert
+# always arrays where the first two indices are atom indices and any further are indexing
+# into the feature if it is itself higher-dimensional. This is not how the weave featurizer
+# currently works...
+
+#=
+This is just an example of another sort of feature one might add, but it's not needed/used
+currently.
+=#
+struct StructureFeat{Tn,Te}<:AbstractFeature{Tn,Te}
+    name
+    encode_f
+    decode_f
+    # etc...
+end
+
+#= FEATURIZATION OBJECTS
+All such objects should define at least one list of <:AbstractFeature objects and a `combine`
+function that specifies how to put them together for eventual attachment to an atoms 
+object. 
+=#
+
+abstract type AbstractFeaturization end
+
+struct GraphNodeFeaturization <: AbstractFeaturization
+    atom_feats::Vector{AtomFeat}
+    feature_vectors::Dict{String,Vector{Float32}} # map from element symbol to vector
+    combine
+end
+
+struct WeaveFeaturization <: AbstractFeaturization
+    atom_feats::Vector{AtomFeat}
+    pair_feats::Vector{PairFeat}
+    combine
+end
+
+
+#= ATOMS OBJECTS
+
+AtomGraph shouldn't have to change too much from its current incarnation. WeaveMol is 
+not currently a thing that can be directly fed into a model, so that will have to get
+updated. Both will need some way to specify which types of features can be attached 
+to them, maybe this is a place for the holy trait design pattern?
+
+Example: AtomGraph can take ElementFeat and ComputedAtomFeat but not PairFeat or StructureFeat, WeaveMol can take ElementFeat, ComputedAtomFeat, and PairFeat but also not StructureFeat
+=#
+
+abstract type AbstractAtoms end
+
+# it is kind of cool to subtype this from LightGraphs so all that machinery "just works"
+# that being said, all of it would "just work" by just pulling out ag.graph so may be
+# more cool than actually practical, hence I've done the abstract type here
+mutable struct AtomGraph <: AbstractAtoms
+    graph::SimpleWeightedGraph{Int32,Float32}
```

Comment:
Why is this limited to 32 bits? Should `AbstractAtoms` parametrise the eltype to constrain things automatically?

Reply:
It is to maintain compatibility with Flux apparently (as mentioned [here](https://github.com/aced-differentiate/ChemistryFeaturization.jl/wiki/The-AtomGraph-Type)).
Having it not limited to 32 bits (if possible) would be nice, and it would make things easier for people adding edges manually into an `AtomGraph`'s graph too.

Reply:
I think we can drop this. It's the flux default, but it's not particularly enforced. In fact, if you find cases where it is, do file an issue and pr.